sdks/python/apache_beam/metrics/cells.py

```diff
@@ -0,0 +1,229 @@
+    with self._lock:
+      self.value += n
+      self.commit.modified()
+
+  def get_cumulative(self):
+    with self._lock:
+      return self.value
+
+
+class DistributionCell(Distribution, MetricCell):
+  """Keeps the state of a distribution metric during pipeline execution.
+
+  It's thread safe.
+  """
+  def __init__(self, *args):
+    super(DistributionCell, self).__init__(*args)
+    self.data = DistributionData(0, 0, None, None)
+
+  def combine(self, other):
+    result = DistributionCell()
+    result.data = self.data.combine(other.data)
+    return result
+
+  def update(self, value):
+    with self._lock:
+      self.commit.modified()
+      self._update(value)
+
+  def _update(self, value):
+    self.data._count += 1
+    self.data._sum += value
+    self.data._min = (value
+                      if self.data.min is None or self.data.min > value
+                      else self.data.min)
+    self.data._max = (value
+                      if self.data.max is None or self.data.max < value
+                      else self.data.max)
+
+  def get_cumulative(self):
+    with self._lock:
+      return self.data.get_cumulative()
+
+
+class DistributionData(object):
+  """The data structure that holds data about a distribution metric.
+
+  This object is not thread safe, so it's not supposed to be modified
+  by other than the DistributionCell that contains it.
+  """
+  def __init__(self, sum, count, min, max):
+    self._sum = sum
+    self._count = count
+    self._min = min
+    self._max = max
+
+  def __eq__(self, other):
+    return (self.sum == other.sum and
+            self.count == other.count and
+            self.min == other.min and
+            self.max == other.max)
+
+  def __neq__(self, other):
+    return not self.__eq__(other)
+
+  def __repr__(self):
+    return '<DistributionData({}, {}, {}, {})>'.format(self.sum,
+                                                       self.count,
+                                                       self.min,
+                                                       self.max)
+
+  def get_cumulative(self):
+    return DistributionData(self.sum, self.count, self.min, self.max)
+
+  @property
+  def sum(self):
+    return self._sum
+
+  @property
+  def count(self):
+    return self._count
+
+  @property
+  def min(self):
+    return self._min
+
+  @property
+  def max(self):
+    return self._max
+
+  @property
+  def mean(self):
+    return self.sum / self.count
+
+  def combine(self, other):
+    if other is None:
+      return self
+    else:
+      new_min = (None if self.min is None and other.min is None else
+                 min(x for x in (self.min, other.min) if x is not None))
+      return DistributionData(
+          self.sum + other.sum,
+          self.count + other.count,
+          new_min,
+          max(self.max, other.max))
+
+  @classmethod
+  def singleton(cls, value):
+    return DistributionData(value, 1, value, value)
+
+
+class MetricAggregation(object):
```

Comment:
Should this be MetricAggregator to be consistent with CounterAggregator and DistributionAggregator below?